Valida ano do CNJ

pipibodock opened the pull request:
Cria validação de ano, para os números de CNJ extraídos.

Fix #4

pipibodock committed:
Adiciona validação de ano ao cnj extraido

- Adiciona método para validação do ano, presente no CNJ extraído
- Corrige typo "aprecem" para "aparecem"

pipibodock committed:
Adiciona limite inferior para validação de ano

- Adiciona limite inferior para a validação feita no ano do CNJ.
Define o limite inferior, como o ano do primeiro processo ajuizado no
Brasil
- Melhora nomes de variáveis

mbodock commented on cnjfacil/extrator.py:
```diff
@@ -35,9 +36,20 @@ def cnjs(self):
+    def _valida_ano_do_cnj(self):
+        limite_inferior = 1895  # Ano do primeiro processo ajuizado do Brasil
+        limite_superior = datetime.datetime.utcnow().year
+        cnjs_invalidos = filter(
```
Que tal invés de gerarmos uma lista com os CNJs a remover, não invertemos a lógica dessa lambda e já filtramos as válidas?

Assim evitamos esse `for` com `list.remove`. Que é O(n²).

Afinal você percorre todos os inválidos, e para cada invalido você percorre a lista o buscando.

pipibodock replied:
Alteração feita. Realmente fez muito mais sentido e menos complexo.
Obrigado pela sugestão.

mbodock commented on cnjfacil/testes/extrator.py:
```diff
@@ -56,10 +56,23 @@ def test_cnj_incorrigivel(self):
         self.assertEqual(extrator.cnjs, ['0053087-35.2013.8.13.0693',
                                          '0516710-11.2017.8.13.0000'])
+
+    def test_nao_adiciona_valores_com_ano_do_cnj_invalido(self):
```
Seria também interessante reduzir a sample desse test. Não precisamos de tanto texto aqui.

Deixe apenas os CNJs separados por uma pequena quantidade de testes para validar sua implementação.

Isso facilita para quem for entender seu teste no futuro.

pipibodock replied:
Para o caso proposto no teste, o tamanho do texto realmente não faz diferença.
O texto foi reduzido.
Obrigado.

mbodock reviewed (changes requested):
Muito obrigado @pipibodock por essa melhoria. Apenas alguns pequenos detalhes a serem corrigidos.

Por favor me avise se precisar de ajuda.

pipibodock committed:
Corrige complexidade da extração de cnj

- Corrige complexidade invertendo a lógica da extração dos cnjs validos
- Diminui texto usado como fixture do teste de validação do ano

mbodock commented:
Merged! 

Muito obrigado @pipibodock !

gagustavo commented on cnjfacil/extrator.py:
```diff
@@ -35,9 +36,18 @@ def cnjs(self):
+        limite_inferior = 1895  # Ano do primeiro processo ajuizado do Brasil
+        limite_superior = datetime.datetime.utcnow().year
+        self._cnjs = filter(
+            lambda cnj: int(cnj[11::][:4]) < limite_superior and int(cnj[11::][:4]) > limite_inferior,
```
@pipibodock esta condição do `limite_superior` deixou de fora o ano atual, no caso, 2020. O correto seria usar o `<=` no limite superior e `>=` no limite inferior (este com menor relevância, por não existir muitos processos deste ano inicial)

pipibodock replied:
Muito obrigado pela correção, @gagustavo.
